### PR TITLE
Use C99 vsnprintf in BIO routines

### PR DESCRIPTION
**Pull request by richsalz**
Replace the formatted print routine in BIO_print
with small wrappers around the C99 functions.
Remove a test, rely on standard documented behavior.

This is a higher-leverage fix than the old #132.  It removes 160 lines and preserves the function and semantics. It's also a lot simpler to verify :)

## Checklist
Thank you for your contribution. Please answer the following:

- [X] I acknowledge that I am authorized to submit this code under
the terms of the [Apache License](https://www.apache.org/licenses/LICENSE-2.0)
- [ ] All new public APIs are documented
- [ ] All new public functions have tests
- [ ] All new public functions have fuzzing tests

### CODE REVIEW

**Review by wbl — DISMISSED** [review]

File: crypto/bio/bio_print.c [hunk 1]
```diff
@@ -8,947 +8,70 @@
...
 {
     va_list args;
-    int ret;
+    size_t ret;
 
     va_start(args, format);
-
-    ret = BIO_vsnprintf(buf, n, format, args);
-
+    ret = vsnprintf(buff, n, format, args);
```
Comment (botovq) [inline crypto/bio/bio_print.c:62]: vsnprintf doesn't return a size_t.

Comment (richsalz) [reply crypto/bio/bio_print.c:63]: aargh.  how did I miss that?  thanks, fixup commit pushed (already built cleanly on my machine)

Comment (botovq) [reply crypto/bio/bio_print.c:63]: BIO_snprintf() and BIO_vsnprintf() still have size_t

I'm really not a fan of these wrappers. As we have seen in the last pull request they also [hide bugs](https://github.com/quictls/quictls/pull/132#discussion_r1871244988) since it's annoying to add the relevant compiler attributes portably.

Comment (richsalz) [reply crypto/bio/bio_print.c:63]: I am not a fan of them either. A major problem is that the BIO functions have a different return value, -1 on error rather than the needed size. Eventually we should clean up our code.

**Review by wbl — APPROVED** [review]

File: crypto/bio/bio_print.c [hunk 2]
```diff
@@ -8,947 +8,71 @@
...
- * function should be renamed, but to what?)
+ * This is a wrapper around snprintf, but returns -1 if the
+ * buffer is too small (documented behavior).
  */
-int BIO_snprintf(char *buf, size_t n, const char *format, ...)
+int BIO_snprintf(char *buff, size_t n, const char *format, ...)
 {
     va_list args;
-    int ret;
+    size_t ret;
```
Comment (botovq) [inline crypto/bio/bio_print.c:60]:
```suggestion
    int ret;
```

File: crypto/bio/bio_print.c [hunk 3]
```diff
@@ -8,947 +8,71 @@
...
+ * buffer is too small (documented).
+ */
+int BIO_vsnprintf(char *buff, size_t n, const char *format, va_list args)
 {
-    size_t retlen;
-    int truncated;
-
-    if (!_dopr(&buf, NULL, &n, &retlen, &truncated, format, args))
-        return -1;
+    size_t ret;
```
Comment (botovq) [inline crypto/bio/bio_print.c:74]:
```suggestion
    int ret;
```

**Review by botovq — CHANGES_REQUESTED** [review]

**Comment by botovq** [comment]
> The change from size_t ret to int ret meant I had to add two casts.

Why? Are you compiling with `-Werror -Wsign-compare` or some sch? The casts don't really do anything by the integer promotion rules.

File: crypto/bio/bio_print.c [hunk 4]
```diff
@@ -8,947 +8,71 @@
...
+	/* Allocate enough space and try again. */
+	int size = ret + 1;
+	char *p = OPENSSL_malloc(size);
+
+	if (p != NULL) {
+	    ret = vsnprintf(p, size, format, copy);
+	    if (ret < size)
+		i = BIO_write(bio, p, ret);
+	    OPENSSL_free(p);
+	}
```
Comment (botovq) [inline crypto/bio/bio_print.c:47]: There's a bunch of tabs that snuck in here, but since you're going to move to a proper code formatter, I'm going to ignore this.

File: test/cmp_ctx_test.c [hunk 5]
```diff
@@ -137,7 +137,8 @@ static int msg_total_size_log_cb(const char *func, const char *file, int line,
                                  OSSL_CMP_severity level, const char *msg)
 {
     msg_total_size += strlen(msg);
-    TEST_note("total=%d len=%zu msg='%s'\n", msg_total_size, strlen(msg), msg);
+    TEST_note("total=%d len=%d msg='%s'\n",
+	    msg_total_size, (int)strlen(msg), msg);
```
Comment (botovq) [inline test/cmp_ctx_test.c:141]: Hrm. Why this change?

**Review by botovq — APPROVED** [review]
I'm reluctantly ok with this, but see two small notes

**Comment by botovq** [comment]
> The :S is for sarcasm? :)

Half smile, half frown. :)

File: crypto/bio/bio_print.c (hunk 4 above)
Comment (richsalz) [reply crypto/bio/bio_print.c:47]: I gotta figure out what went wrong, why `:set et` didn't work.

File: test/cmp_ctx_test.c (hunk 5 above)
Comment (richsalz) [reply test/cmp_ctx_test.c:141]: I don't know why I thought `%zu` wasn't standard.  Might as well undo that change and fix the tab issue at the same time.

File: crypto/bio/bio_print.c (hunk 4 above)
Comment (botovq) [reply crypto/bio/bio_print.c:47]: What often bites me is that it's local to the buffer.

Comment (richsalz) [reply crypto/bio/bio_print.c:47]: I had it set for openssl not quictls :(. fixed.

**Comment by richsalz** [comment]
@wahern, those all seem like valid points.  Please make a PR.
